README.md
# folded/crypt

Easily encrypt and decrypt strings for your web app.

[![Build Status](https://travis-ci.com/folded-php/crypt.svg?branch=master)](https://travis-ci.com/folded-php/crypt) [![Maintainability](https://api.codeclimate.com/v1/badges/8e59f7e6108adc65ca6b/maintainability)](https://codeclimate.com/github/folded-php/crypt/maintainability) [![TODOs](https://img.shields.io/endpoint?url=https://api.tickgit.com/badge?repo=github.com/folded-php/crypt)](https://www.tickgit.com/browse?repo=github.com/folded-php/crypt)

## Summary

- [About](#about)
- [Features](#features)
- [Requirements](#requirements)
- [Installation](#installation)
- [Examples](#examples)
- [Version support](#version-support)

## About

I created this library to be able to encrypt my data in a standalone way.

Folded is a constellation of packages to help you setting up a web app easily, using ready to plug in packages.

- [folded/action](https://github.com/folded-php/action): A way to organize your controllers for your web app.
- [folded/config](https://github.com/folded-php/config): Configuration utilities for your PHP web app.
- [folded/exception](https://github.com/folded-php/exception): Various kind of exception to throw for your web app.
- [folded/history](https://github.com/folded-php/history): Manipulate the browser history for your web app.
- [folded/http](https://github.com/folded-php/http): HTTP utilities for your web app.
- [folded/orm](https://github.com/folded-php/orm): An ORM for you web app.
- [folded/request](https://github.com/folded-php/request): Request utilities, including a request validator, for your PHP web app.
- [folded/routing](https://github.com/folded-php/routing): Routing functions for your PHP web app.
- [folded/session](https://github.com/folded-php/session): Session functions for your web app.
- [folded/view](https://github.com/folded-php/view): View utilities for your PHP web app.

## Features

- Can encrypt and decrypt strings
- Can generate a key (necessary to setup the library) from the command line

## Requirements

- PHP >= 7.4.0
- Composer installed

## Installation

- [1. Install the package](#1-install-the-package)
- [2. Generate a key](#2-generate-a-key)
- [3. Add the setup code](#3-add-the-setup-code)

### 1. Install the package

In your root folder, run this command:

```bash
composer require folded/crypt
```

### 2. Generate a key

One way to generate the key easily is through the command line. Run this command to get a new key:

```bash
vendor/bin/crypt generate key
```

You can get more information on the available option by running `vendor/bin/crypt generate --help`.

Another way is to call the function `Folded\getEncryptionKey()` from a script:

```php
use function Folded\getEncryptionKey;

require __DIR__ . "/vendor/autoload.php";

echo getEncryptionKey();
```

You can add a parameter to control the type of cipher you want (currently supported: AES-128-CBC and AES-256-CBC).

### 3. Add the setup code

Before calling the library, add this setup code as early as possible:

```php
use function Folded\setEncryptionKey;

setEncryptionKey("xIYrZSsCV6hx9x/Q4bka1PejU+aSaMerJQFSYr3QnTE=");
```

## Examples

- [1. Encrypt a string](#1-encrypt-a-string)
- [2. Decrypt a string](#2-decrypt-a-string)

### 1. Encrypt a string

In this example, we will get the encrypted version of a string.

```php
use function Folded\getEncryptedString;

$encryptedText = getEncryptedString("hello world");
```

### 2. Decrypt a string

In this example, we will decrypt a previously encrypted text.

```php
use function Folded\getDecryptedString;

$encryptedString = "...";
$decryptedString = getDecryptedString($encryptedString);
```

Note it will only decrypt encrypted string from the `getEncryptedString()` function.

Also note that if you encrypt a string with a key A, and you change the key A for a new key, the `getDecryptedString` will not be able to successfuly decrypt the text and get the original text so be careful to save your key in somewhere safe (generally in a .env file).

## Version support

|        | 7.3 | 7.4 | 8.0 |
| ------ | --- | --- | --- |
| v0.1.0 | ❌  | ✔️  | ❓  |


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.1] 2020-10-01

### Fix

- Bug when the `function_exists` statements were missing the namespace `Folded`.

## [0.1.0] 2020-09-17

### Added

- First working version.


tests/CryptTest.php
<?php

declare(strict_types = 1);

use Folded\Crypt;

beforeEach(function (): void {
    Crypt::clear();
});

it("should decrypt the encrypted string", function (): void {
    $data = "Hello world";

    Crypt::setKey(Crypt::getKey());

    $encryptedString = Crypt::getEncryptedString($data);

    expect(Crypt::getDecryptedString($encryptedString))->toBe($data);
});

it("should throw an exception if the key has not been set", function (): void {
    $this->expectException(Exception::class);

    Crypt::getEncryptedString("foo");
});

it("should throw an exception message if the key has not been set", function (): void {
    $this->expectExceptionMessage("key has not been set");

    Crypt::getEncryptedString("foo");
});

it("should throw an exception if the cipher is empty", function (): void {
    $this->expectException(InvalidArgumentException::class);

    Crypt::setKey("foo", "");
});

it("should throw an exception message if the cipher is empty", function (): void {
    $this->expectExceptionMessage("cipher is empty");

    Crypt::setKey("foo", "");
});

it("should throw an exception if the key is empty", function (): void {
    $this->expectException(InvalidArgumentException::class);

    Crypt::setKey("");
});

it("should throw an exception message if the key is empty", function (): void {
    $this->expectExceptionMessage("key is empty");

    Crypt::setKey("");
});

it("should throw an exception if the cipher is not correct", function (): void {
    $this->expectException(InvalidArgumentException::class);

    Crypt::setKey("foo", "bar");
});

it("should throw an exception message if the cipher is not correct", function (): void {
    $this->expectExceptionMessage("cipher bar not allowed (allowed: AES-128-CBC, AES-256-CBC)");

    Crypt::setKey("foo", "bar");
});


tests/getEncryptionKeyTest.php
<?php

declare(strict_types = 1);

use Folded\Crypt;
use function Folded\getEncryptionKey;

beforeEach(function (): void {
    Crypt::clear();
});

it("should return a random string", function (): void {
    expect(getEncryptionKey())->toBeString();
});


tests/setEncryptionKeyTest.php
<?php

declare(strict_types = 1);

use Folded\Crypt;
use function Folded\getEncryptionKey;
use function Folded\setEncryptionKey;

beforeEach(function (): void {
    Crypt::clear();
});

it("should return null", function (): void {
    expect(setEncryptionKey(getEncryptionKey()))->toBeNull();
});


src/Crypt.php
<?php

declare(strict_types = 1);

namespace Folded;

use Exception;
use Illuminate\Encryption\Encrypter;
use InvalidArgumentException;

/**
 * Encrypts and decrypts strings.
 *
 * @since 0.1.0
 */
class Crypt
{
    /**
     * @since 0.1.0
     */
    const ALLOWED_CIPHERS = [
        "AES-128-CBC",
        "AES-256-CBC",
    ];

    /**
     * @since 0.1.0
     */
    const DEFAULT_CIPHER = "AES-256-CBC";

    /**
     * The cipher to use for the key.
     *
     * @since 0.1.0
     */
    private static string $cipher = self::DEFAULT_CIPHER;

    /**
     * The engine that will encrypt and decrypt strings.
     *
     * @since 0.1.0
     */
    private static ?Encrypter $engine = null;

    /**
     * The encryption key used as a salt to improve the security of encrypted strings.
     *
     * @since 0.1.0
     */
    private static string $key = "";

    /**
     * Clear the state.
     * Useful for unit testing.
     *
     * @since 0.1.0
     *
     * @example
     * Crypt::clear();
     */
    public static function clear(): void
    {
        self::$key = "";
        self::$cipher = "";
        self::$engine = null;
    }

    /**
     * Get the decrypted string from an encrypted string.
     *
     * @param string $cryptedString The encrypted string.
     *
     * @throws Exception If the key has not been set.
     *
     * @since 0.1.0
     *
     * @example
     * Crypt::getDecryptedString("...");
     */
    public static function getDecryptedString(string $cryptedString): string
    {
        self::checkKeySet();

        return self::engine()->decryptString($cryptedString);
    }

    /**
     * Encrypts a strings.
     *
     * @param string $data The text to encrypt.
     *
     * @throws Exception If the key has not been set.
     *
     * @since 0.1.0
     *
     * @example
     * Crypt::getEncryptedString("foo");
     */
    public static function getEncryptedString(string $data): string
    {
        self::checkKeySet();

        return self::engine()->encryptString($data);
    }

    /**
     * Get a random secure key to be used to encrypt strings with this library.
     *
     * @param string $cipher The cipher to use (default: AES-256-CBC).
     *
     * @since 0.1.0
     *
     * @example
     * Crypt::getKey();
     */
    public static function getKey(string $cipher = self::DEFAULT_CIPHER): string
    {
        self::checkCipher($cipher);

        return base64_encode(Encrypter::generateKey($cipher));
    }

    /**
     * Set the key that is used as a salt to improve the security of the encrypted strings.
     *
     * @param string $key    The key to be used as a salt.
     * @param string $cipher The cipher algorithm to use (default: AES-256-CBC).
     *
     * @throws InvalidArgumentException If the cipher is incorrect.
     * @throws InvalidArgumentException If the cipher is empty.
     * @throws InvalidArgumentException If the key is empty.
     *
     * @since 0.1.0
     *
     * @example
     * Crypt::setKey("...");
     */
    public static function setKey(string $key, string $cipher = self::DEFAULT_CIPHER): void
    {
        if (empty(trim($key))) {
            throw new InvalidArgumentException("key is empty");
        }

        if (empty(trim($cipher))) {
            throw new InvalidArgumentException("cipher is empty");
        }

        self::checkCipher($cipher);

        self::$key = base64_decode($key, true);
        self::$cipher = $cipher;
    }

    /**
     * Throws an exception if the cipher is incorrect.
     *
     * @throws InvalidArgumentException If the cipher is not correct.
     *
     * @since 0.1.0
     *
     * @example
     * Crypt::checkCipher();
     */
    private static function checkCipher(string $cipher): void
    {
        if (!in_array($cipher, self::ALLOWED_CIPHERS, true)) {
            throw new InvalidArgumentException("cipher $cipher not allowed (allowed: " . implode(", ", self::ALLOWED_CIPHERS) . ")");
        }
    }

    /**
     * Throws an exception if the key is not correct.
     *
     * @throws Exception If the key is not set.
     *
     * @since 0.1.0
     *
     * @example
     * Crypt::checkKeySet();
     */
    private static function checkKeySet(): void
    {
        if (empty(trim(self::$key))) {
            throw new Exception("key has not been set");
        }
    }

    /**
     * Returns an engine that encrypts and decrypts strings (acting as a singleton).
     *
     * @since 0.1.0
     *
     * @example
     * $engine = Crypt::engine();
     */
    private static function engine(): Encrypter
    {
        if (!(self::$engine instanceof Encrypter)) {
            self::$engine = new Encrypter(self::$key, self::$cipher);
        }

        return self::$engine;
    }
}
